napari/util/keybindings.py

```diff
@@ -0,0 +1,582 @@
+def components_to_seq(key, modifiers):
+    """Combine components to become a key sequence.
+
+    Modifier keys will always be combined in the same order:
+    Control, Alt, Shift, Meta
+
+    Shift pressed with any letter will be consumed
+    to transform it to upper case, e.g. Shift + w => W.
+    Otherwise, the letter will be lower case.
+    Shift will also be consumed without transformation when pressed
+    with any other non-special key unless Control is also a modifier.
+
+    Parameters
+    ----------
+    key : str or vispy.app.Key
+        Base key.
+    modifiers : sequence of str or vispy.app.Key
+        Modifier keys.
+
+    Returns
+    -------
+    seq : str
+        Generated key sequence.
+    """
+    if len(key) == 1 and key.isalpha():  # it's a letter
+        if 'Shift' not in modifiers:
+            key = key.lower()
+            cond = lambda m: True  # noqa: E731
+        else:
+            # Shift is consumed to create upper-case letter
+            key = key.upper()
+            cond = lambda m: m != 'Shift'  # noqa: E731
+    elif key in SPECIAL_KEYS:
+        # remove redundant information i.e. an output of 'Shift-Shift'
+        cond = lambda m: m != key  # noqa: E731
+    else:
+        # Shift is consumed to transform key
+
+        # bug found on OSX: Command will cause Shift to not
+        # transform the key so do not consume it
+        # note: 'Control' is OSX Command key
+        cond = lambda m: m != 'Shift' or 'Control' in modifiers  # noqa: E731
+
+    modifiers = tuple(key.name for key in
+                      filter(lambda key: key in modifiers and cond(key),
+                             MODIFIER_KEYS.values()))
+
+    return '-'.join(modifiers + (key,))
+
+
+def normalize_key_sequence(seq):
+    """Normalize key sequence to make it easily comparable.
+
+    All aliases are converted and modifier orders are fixed to:
+    Control, Alt, Shift, Meta
+
+    Parameters
+    ----------
+    seq : str
+        Key sequence.
+
+    Shift pressed with any letter will be consumed
+    to transform it to upper case, e.g. Shift + w => W.
+    Otherwise, the letter will be lower case.
+    Shift will also be consumed without transformation when pressed
+    with any other non-special key unless Control is also a modifier.
+
+    Returns
+    -------
+    normalized_seq : str
+        Normalized key sequence.
+    """
+    seq = _expand_shorthand(seq)
+
+    key, modifiers = parse_seq(seq)
+
+    if len(key) != 1 and key not in SPECIAL_KEYS:
+        raise TypeError(f'invalid key {key}')
+
+    for modifier in modifiers:
+        if modifier not in MODIFIER_KEYS.values():
+            raise TypeError(f'invalid modifier key {modifier}')
+
+    return components_to_seq(key, modifiers)
+
+
+def _determine_binding_op(arg):
+    """Determine the binding operation based on the argument passed.
+
+    None     -> BindOperation.UNBIND
+    str      -> BindOperation.REBIND
+    callable -> BindOperation.BIND
+    """
```

Comment:
Can you elaborate on why this function exists? Where is it used? (I know I can go and look for usage but in general *code should be local*, ie you should not need to have global context in your head to understand the piece of code you're reading.